### PR TITLE
Add error message to account registration error

### LINKED ISSUE #9231: `certbot register` is swallowing error details
image

Certbot is swallowing the registration error. 

One way to reproduce this is to configure Pebble to require EAB for registration. The underlying error is nowhere to be seen, unless you delve into the log files:

    certbot.errors.Error: Server requires external account binding. Please use --eab-kid and --eab-hmac-key.

### PR DESCRIPTION
**Pull request by osirisinferi**
Fixes #9231

I've opted to have the first character of the error message to be lower case, as it's the first character after a colon. For me personally, that has to be lower case instead of upper case.

### CODE REVIEW

**Review by alexzorin — CHANGES_REQUESTED** [review]
Thanks for picking this up.

>I've opted to have the first character of the error message to be lower case, as it's the first character after a colon. For me personally, that has to be lower case instead of upper case.

I think if we choose to do this, we should do it consistently across Certbot's error handling. At the moment, it looks like all cases of errors following a colon, show the error verbatim.

I am also a little skeptical that it won't sometimes do the wrong thing. e.g. if the first word in the error message is `ACME`, you would get `aCME`, which is weirdly reminiscent of ASN.1 ;).

**Comment by osirisinferi** [comment]
Note that the error "Server requires external account binding. (...)" is actually a message from `perform_registration()` in `client.py` and not actually an error message from the ACME API.

**Comment by osirisinferi** [comment]
Code is a little bit more complicated now to also catch errors from the acme library. It raises `acme.messages.Error` and it contains the `str` `detail` if provided by the ACME API. If not, the acme library can return a human readable description of the ACME error type using the `description` property.

An example would be an `urn:ietf:params:acme:error:serverInternal` error.

**Comment by osirisinferi** [comment]
By the way, that whole part of `_determine_account()` doesn't have tests and I didn't add some too.. py3-cover doesn't seem to complain though.. :roll_eyes:

File: certbot/certbot/_internal/main.py [hunk 1]
```diff
@@ -726,10 +727,28 @@ def _tos_cb(terms_of_service: str) -> None:
...
+                    elif err.description:
+                        # If `detail` is not provided, the `acme` library can return a human
+                        # readable error message based on the error type in `description`.
+                        err_msg = err.description
+                    else:
+                        # If no human readable error message can be found, the ACME error type
+                        # is presented to the user.
+                        err_msg = str(err)
+                else:
+                    err_msg = str(err)
```
Comment (alexzorin) [inline certbot/certbot/_internal/main.py:748]: Sorry for the delay reviewing this, it's been hard to find time.

For consistency and simplicity's sake, ISTM that we can just take advantage of `__str__` on `acme.messages.Error`:

https://github.com/certbot/certbot/blob/142fcad28b0e61f335b92a743b38c8d677bd00dd/acme/acme/messages.py#L137-L141

Would you be able to give that a try and see whether `str(err_msg)` does the trick? I'd like to avoid having one-off code related to display here. If we do end up needing a new way to display errors to users, we should deal with it in `acme`.

Comment (osirisinferi) [reply certbot/certbot/_internal/main.py:748]: I noticed that function too when looking at it. But if we use `str(err)` it would mean a Certbot raised error would be a relative simple error message which would be *very* different than an error message raised by the ACME client, as that would be a long line with all the different fields separated by " :: ". To me, that difference was too big, so I build the code above to make it more the same with regard to a user experience: a simple to understand error message.

But of course [KISS](https://en.wikipedia.org/wiki/KISS_principle) is always a good argument :wink:

Comment (alexzorin) [reply certbot/certbot/_internal/main.py:748]: That's reasonable, and there's always the option to change what `__str__` does, if we do decide it's unreadable.

https://datatracker.ietf.org/doc/html/rfc7807#section-3.1 is pretty clear that only `title` and `detail` are supposed to be human-readable.

I also recall reading a comment on the Certbot repo years ago from one of Boulder's developers, calling Certbot out for the way the `description` property is interpreting the problem `type` and mapping it into [this dictionary](https://github.com/certbot/certbot/blob/142fcad28b0e61f335b92a743b38c8d677bd00dd/acme/acme/messages.py#L35-L66). This was a misinterpretation of the ACME document, I think.

tl;dr; I'm sure we'd be open to changing it, but let's kick it to another issue.

Comment (osirisinferi) [reply certbot/certbot/_internal/main.py:748]: I'll make this code as simple as possible, so we can change anything, if required, from a single point in `__str__()` in the future.

**Comment by osirisinferi** [comment]
I did leave the distinction between a Certbot generated error and ACME server generated error and I removed the "Error returned by the ACME server: " part for a Certbot generated error, as that was incorrect before.

Also note that the contents `test_register_error_acme_type_and_detail()` test would never be presented to the user by Certbot, as Certbot would have raised its own error when the user chooses *not* to agree with the ToS.. But I think it's worth having a test for that specific case anyway, the actual contents don't really matter in that regard I think.

File: certbot/certbot/_internal/main.py [hunk 2]
```diff
@@ -726,10 +727,14 @@ def _tos_cb(terms_of_service: str) -> None:
                 display_util.notify("Account registered.")
             except errors.MissingCommandlineFlag:
                 raise
-            except errors.Error:
+            except (errors.Error, acme_messages.Error) as err:
                 logger.debug("", exc_info=True)
+                if isinstance(err, acme_messages.Error):
```
Comment (alexzorin) [inline certbot/certbot/_internal/main.py:732]:
```suggestion
                if acme_messages.is_acme_error(err):
```

File: certbot/tests/main_test.py [hunk 3]
```diff
@@ -593,6 +594,17 @@ def _call(self):
             mock_storage.return_value = self.account_storage
             return _determine_account(self.config)
 
+    @mock.patch('certbot._internal.client.register')
+    @mock.patch('certbot._internal.client.display_ops.get_email')
+    def _register_error_base(self, err_msg, exception, mock_get_email, mock_register):
```
Comment (alexzorin) [inline certbot/tests/main_test.py:599]: To follow existing conventions:

```suggestion
    def _register_error_common(self, err_msg, exception, mock_get_email, mock_register):
```

File: certbot/tests/main_test.py [hunk 4]
```diff
@@ -641,6 +663,24 @@ def test_no_accounts_email(self):
...
+        err_msg = "Some error message raised by Certbot"
+        self._register_error_base(err_msg, errors.Error(err_msg))
+
+    def test_register_error_acme_type_and_detail(self):
+        err_msg = ("Error returned by the ACME server: urn:ietf:params:acme:"
+                   "error:malformed :: The request message was malformed :: "
+                   "must agree to terms of service")
+        exception = acme_error(typ = "urn:ietf:params:acme:error:malformed",
+                               detail = "must agree to terms of service")
+        self._register_error_base(err_msg, exception)
```
Comment (alexzorin) [inline certbot/tests/main_test.py:676]:
```suggestion
        self._register_error_common(err_msg, exception)
```

File: certbot/tests/main_test.py [hunk 5]
```diff
@@ -641,6 +663,24 @@ def test_no_accounts_email(self):
...
+
+    def test_register_error_acme_type_only(self):
+        err_msg = ("Error returned by the ACME server: urn:ietf:params:acme:"
+                   "error:serverInternal :: The server experienced an internal error")
+        exception = acme_error(typ = "urn:ietf:params:acme:error:serverInternal")
+        self._register_error_base(err_msg, exception)
```
Comment (alexzorin) [inline certbot/tests/main_test.py:682]:
```suggestion
        self._register_error_common(err_msg, exception)
```

File: certbot/tests/main_test.py [hunk 6]
```diff
@@ -641,6 +663,24 @@ def test_no_accounts_email(self):
         self.assertEqual(self.accs[1].id, self.config.account)
         self.assertEqual('other email', self.config.email)
 
+    def test_register_error_certbot(self):
+        err_msg = "Some error message raised by Certbot"
+        self._register_error_base(err_msg, errors.Error(err_msg))
```
Comment (alexzorin) [inline certbot/tests/main_test.py:668]:
```suggestion
        self._register_error_common(err_msg, errors.Error(err_msg))
```

File: certbot/certbot/_internal/main.py [hunk 7]
```diff
@@ -726,10 +727,14 @@ def _tos_cb(terms_of_service: str) -> None:
...
+                    err_msg = f"Error returned by the ACME server: {str(err)}"
```
Comment (alexzorin) [inline certbot/certbot/_internal/main.py:733]: Lol:

https://github.com/certbot/certbot/blob/142fcad28b0e61f335b92a743b38c8d677bd00dd/certbot/certbot/_internal/log.py#L367-L370

I am starting to wonder we should change this properly, now. I'm actually not sure whether messing with `__str__` would be a breaking change, but we can always add a new method for display purposes. Let me find out.

**Review by alexzorin — CHANGES_REQUESTED** [review]

File: certbot/certbot/_internal/main.py (hunk 7 above)
Comment (osirisinferi) [reply certbot/certbot/_internal/main.py:733]: Hm, that code *assumes* that an ACME error *always* contains all three properties. But in `acme.messages.Error` all three properties have `omitempty=True` set. So if the ACME server doesn't e.g. provide the title for some reason, `str(err)` would only provide two partitions separated by `:: `, right? Not sure if Boulder actually does that and not sure if it's allowed by the RFCs, but still.

Also, not sure how it's tied into this PR.

Comment (alexzorin) [reply certbot/certbot/_internal/main.py:733]: I would really like to not show the error type and the description field, but let's leave it for now. This PR fixes the problem it was supposed to, albeit slightly verbosely.

**Review by alexzorin — APPROVED** [review]
